### PR TITLE
Introduce Neighbors

### PR DESCRIPTION
**Pull request by ajdapretnar**
##### Issue

Move OWNeighbors from Prototypes to core.

##### Description of changes
Neighbors widget moved to core. Tests are fixed (before they weren't working). Documentation added and updated (to an extent).

##### Includes
- [X] Code changes
- [X] Tests
- [X] Documentation

### CODE REVIEW

**Comment by ajdapretnar** [comment]
@janezd I am not sure what to do regarding the changes you proposed. I have simply moved an existing widget to core. I think the new changes should probably be done by the person who designed Neighbors, but I can give it a go if you want.

**Comment by janezd** [comment]
Comments were not meant for you but for whoever would nurture this widget into the core. I can live with `apply` as it is, but I'd like to see new-style signals (because hygiene), `handleNewSignals` (because potential problems) and checkbox to the left of the label (because conventions).

You can do this, or you can assign it to somebody (@VesnaT is the original author). If `apply` is not changed, you can write an issue with a reference to my comments.

File: Orange/widgets/data/owneighbors.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        self.apply()
+
+    def apply(self):
+        if self.data is None or self.reference is None:
+            self.send("Neighbors", None)
+            return
+        distance = METRICS[self.distance_index][1]
+        n_data, n_ref = len(self.data), len(self.reference)
+        all_data = Table.concatenate([self.reference, self.data], 0)
+        pp_all_data = Impute()(RemoveNaNColumns()(all_data))
```
[the remark this answers is not in the document]
Comment (ajdapretnar) [reply Orange/widgets/data/owneighbors.py]: This might be the case.

**Comment by VesnaT** [comment]
No worries, I can do it... unless you want to.

**Comment by ajdapretnar** [comment]
I approve the changes. Everything works well.

File: doc/visual-programming/source/widgets/data/neighbors.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        A data table of nearest neighbors according to reference.
+
+
+The **Neighbors** widget computes nearest neighbors for a given reference and for a given distance measure.
+
+.. figure:: images/Neighbors-stamped.png
+   :scale: 50%
+
+1. Information on the input data.
+2. Distance measure for computing neighbors. Supported measures are: Euclidean, Manhattan, Mahalanobis, Cosine, Jaccard, Spearman, absolute Spearman, Pearson, absolute Pearson. If *Exclude references* is ticked, reference data won't be included in the output.
```
[the remark this answers is not in the document]
Comment (PrimozGodec) [reply doc/visual-programming/source/widgets/data/neighbors.rst]: @janezd Currently, all instances that have distance ~0 are excluded (dist < 1e-5). Should it stay like this?

[the remark this answers is not in the document]
Comment (janezd) [reply doc/visual-programming/source/widgets/data/neighbors.rst]: Yes. All instances that are equal to the reference are treated as reference. This should be explicitly written in documentation though. @ajdapretnar, can you add this?

[the remark this answers is not in the document]
Comment (ajdapretnar) [reply doc/visual-programming/source/widgets/data/neighbors.rst]: Yes.

[the remark this answers is not in the document]
Comment (ajdapretnar) [reply doc/visual-programming/source/widgets/data/neighbors.rst]: @PrimozGodec I have added a line explaining this. Could you please check the explanation is clear enough?

[the remark this answers is not in the document]
Comment (PrimozGodec) [reply doc/visual-programming/source/widgets/data/neighbors.rst]: Thank you @ajdapretnar. It sounds clear enough to me. :)

**Pull request merged**